library/alloc/src/vec/is_zero.rs

```diff
@@ -147,6 +147,24 @@ impl_is_zero_option_of_nonzero!(
     NonZeroIsize,
 );
 
+macro_rules! impl_is_zero_option_of_num {
+    ($($t:ty,)+) => {$(
+        unsafe impl IsZero for Option<$t> {
+            #[inline]
+            fn is_zero(&self) -> bool {
+                core::intrinsics::assert_zero_valid::<Option<$t>>();
```

Comment:
I don't think this assert is sufficient since it doesn't tell us which variant has the zero encoding.

Reply:
AFAIK `assert_zero_valid` is for runtime, but because this a macro with known types are I think we could use CTFE to check what we care about instead.  Something like:
```rust
const {
    let z: Self = unsafe { mem::zeroed() };
    assert!(z.is_none());
}
```
which ensures the thing we particularly care about: that it's legal to initialize memory with zeros in order to get a `None` for this type.